File: speech/windowing.lua
-- Copyright (c) 2017-present, Facebook, Inc.
-- All rights reserved.

-- This source code is licensed under the BSD-style license found in the
-- LICENSE file in the root directory of this source tree. An additional grant
-- of patent rights can be found in the PATENTS file in the same directory.

local speech = require('libspeech')

--N is the size of the input
--Window function needs to take two arguments
--n and N (current sample and max samples)
local function Windowing(window, N, inplace)
   local function hanning (n, N)
      return 0.5*(1 -  math.cos(2 * math.pi * (n - 1) / (N - 1)))
   end
   window = window or hanning
   local coefs = torch.range(1, N)
   coefs:apply(function (n) return window(n, N) end)

   return function (output_raw, input_raw)
      local input, output = speech.Proc(output_raw, input_raw)
      coefs = coefs:type(input:type())
      if input:dim() == 1 then
         if inplace then
            input:cmul(coefs)
         else
            output:resize(input:size(1))
            output:copy(input)
            return output:cmul(coefs)
         end
      elseif input:dim() == 2 then
         if inplace then
            input:cmul(coefs:view(1, N):expandAs(input))
         else
            output:resizeAs(input)
            output:copy(input)
            return output:cmul(coefs:view(1, N):expandAs(input))
         end
      else
         error("Unsupported dimension.")
      end
   end
end

speech.Windowing = Windowing

File: src/libraries/feature/Windowing.cpp
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "Windowing.h"

#include <cmath>
#include <cstddef>
#include <numeric>
#include <stdexcept>

namespace w2l {

Windowing::Windowing(int N, WindowType windowtype)
    : windowLength_(N), windowType_(windowtype), coefs_(N) {
  if (windowLength_ <= 1) {
    throw std::invalid_argument("Windowing: windowLength must be > 1");
  }
  std::iota(coefs_.begin(), coefs_.end(), 0.0);
  switch (windowtype) {
    case WindowType::HAMMING:
      for (auto& c : coefs_) {
        c = 0.54 - 0.46 * std::cos(2 * M_PI * c / (N - 1));
      }
      break;
    case WindowType::HANNING:
      for (auto& c : coefs_) {
        c = 0.5 * (1.0 - std::cos(2 * M_PI * c / (N - 1)));
      }
      break;
    default:
      throw std::invalid_argument("Windowing: unsupported window type");
  }
}

std::vector<float> Windowing::apply(const std::vector<float>& input) const {
  auto output(input);
  applyInPlace(output);
  return output;
}

void Windowing::applyInPlace(std::vector<float>& input) const {
  if (input.size() % windowLength_ != 0) {
    throw std::invalid_argument(
        "Windowing: input size is not divisible by windowLength");
  }
  size_t n = 0;
  for (auto& in : input) {
    in *= coefs_[n++];
    if (n == windowLength_) {
      n = 0;
    }
  }
}
} // namespace w2l
